text/2020-07-22-joint-consensus.md

```diff
@@ -0,0 +1,181 @@
+# Use joint consensus
+
+## Summary
+
+To support safe conf change across data center, we need to use joint consensus.
+This RFC describes how to achieve it.
+
+## Motivation
+
+Current approach of single step is known to be not safe when replacing a peer
+within the same node or same DC. Take (a, b, c) as an example, to replacing c
+by d step by step, it has to add d and then remove c. At some point the
+configuration becomes (a, b, c, d). If c and d are on the same node or in the
+same DC, crash of single node or single DC can make the region unavailable even
+there are still two peers up.
+
+## Detailed design
+
+The design and implementation of joint consensus algorithm is not included in
+this RFC. It should be covered by raft paper and the design in
+etcd-io/etcd#7625. And the implementation has been ported to our fork
+tikv/raft-rs.
+
+This RFC focus on how to use and test raft-rs's support of joint consensus in
+TiKV.
+
+### Protocol
+
+To keep compatibility, we need to define a new command to utilize joint
+consensus. We can name it `ConfChange` or `ChangePeerV2`. The change should
+look like
+
+```protobuf
+message ChangePeerV2Request {
+    repeated ChangePeerRequest changes = 1;
+}
+
+message ChangePeerResponse {
+    metapb.Region region = 1;
+}
+```
+
+Because now that we use two phases to commit a confchange, so we need to store
+joint state. Giving that we already use `is_learner` flag to indicate its role,
+we can reuse the flag. According to [protocol updating rules][1], it's OK to
+change a bool field to an enum field. So we can define the new peer meta as
+follow:
+
+```protobuf
+enum PeerRole {
+    // Voter -> Voter
+    Voter = 0;
+    // Learner/None -> Learner
+    Learner = 1;
+    // Learner/None -> Voter
+    IncomingVoter = 2;
+    // Voter -> Learner
+    DemotingVoter = 3;
+    // We forbid Voter -> None, it can introduce unavailability as discussed in
+    // etcd-io/etcd#7625
+    // Learner -> None can be apply directly, doesn't need to be stored as
+    // joint state.
+}
+
+message Peer {
+    uint64 id = 1;
+    uint64 store_id = 2;
+    PeerRole role = 3;
+}
+```
+
+### PD
+
+PD should not propose ChangePeerV2 until all nodes are upgraded to supporting
+versions.
+
+When using joint consensus to remove a Voter, PD should make it into a learner
+first, and then remove it as discussed in etcd-io/etcd#7625.
```

Comment:
Can we automatically demote voter and then delete it, instead of explicit? BTW,Is it still unable to resolve this issue https://github.com/tikv/raft-rs/issues/192?

Reply:
Yes. This can't be resolved. Raft can't work correctly when losing majority.

Reply:
> Can we automatically demote voter and then delete it

TiKV can't do that reliably. For example, leader change can lose progress.